oxide-auth-db/src/primitives/db_registrar.rs

```diff
@@ -0,0 +1,258 @@
+
+impl DBRegistrar {
+    /// Create an DB connection recording to features.
+    pub fn from_url(url: String, max_pool_size: u32) -> Self {
+        DBRegistrar {
+            repo: DataSource::from_url(url, max_pool_size),
+            password_policy: None,
+        }
+    }
+
+    /// Insert or update the client record.
+    pub fn register_client(&mut self, client: Client) -> Result<(), RegistrarError> {
+        let password_policy = Self::current_policy(&self.password_policy);
+        let encoded_client = client.encode(password_policy);
+
+        self.repo
+            .regist_from_encoded_client(encoded_client)
+            .map_err(|_e| RegistrarError::Unspecified)
+    }
+
+    /// Change how passwords are encoded while stored.
+    pub fn set_password_policy<P: PasswordPolicy + 'static>(&mut self, new_policy: P) {
+        self.password_policy = Some(Box::new(new_policy))
+    }
+
+    // This is not an instance method because it needs to borrow the box but register needs &mut
+    fn current_policy<'a>(policy: &'a Option<Box<dyn PasswordPolicy>>) -> &'a dyn PasswordPolicy {
+        policy
+            .as_ref()
+            .map(|boxed| &**boxed)
+            .unwrap_or(&*DEFAULT_PASSWORD_POLICY)
+    }
+}
+
+impl Extend<Client> for DBRegistrar {
+    fn extend<I>(&mut self, iter: I)
+    where
+        I: IntoIterator<Item = Client>,
+    {
+        iter.into_iter().for_each(|client| {
+            self.register_client(client);
+        })
+    }
+}
+
+impl Registrar for DBRegistrar {
+    fn bound_redirect<'a>(&self, bound: ClientUrl<'a>) -> Result<BoundClient<'a>, RegistrarError> {
+        let client = match self.repo.find_client_by_id(bound.client_id.as_ref()) {
+            Ok(detail) => detail,
+            _ => return Err(RegistrarError::Unspecified),
+        };
+        // Perform exact matching as motivated in the rfc
+        let registered_url = match bound.redirect_uri {
+            None => client.redirect_uri.clone(),
+            Some(ref url) => {
+                let original = std::iter::once(&client.redirect_uri);
+                let alternatives = client.additional_redirect_uris.iter();
+                if let Some(registered) = original
+                    .chain(alternatives)
+                    .find(|&registered| *registered == *url.as_ref())
+                {
+                    registered.clone()
+                } else {
+                    return Err(RegistrarError::Unspecified);
+                }
+            }
+        };
+        Ok(BoundClient {
+            client_id: bound.client_id,
+            redirect_uri: Cow::Owned(registered_url),
+        })
+    }
+
+    fn negotiate<'a>(
+        &self, bound: BoundClient<'a>, _scope: Option<Scope>,
+    ) -> Result<PreGrant, RegistrarError> {
+        let client = self
+            .repo
+            .find_client_by_id(&bound.client_id)
+            .map_err(|_e| RegistrarError::Unspecified)
+            .unwrap();
```

Comment:
`map_err` should be followed by `?` and not `unwrap`.

Reply:
Resolved, thanks again.